### ISSUE TITLE
service: write operations & porcelain

### ISSUE THREAD

**Issue by jsam** [SwissDataScienceCenter/renku-python#1870]
So far all endpoints implemented on the service were designed to achieve most atomic renku operations. This approach is important to achieve some of the initial design goals, one of them being able to write clients in other languages which can be on par with the python client (e.g. R client) and offline mode. On the other hand core service also has stateless clients (i.e. UI), which currently struggle to maintain states and would like to push state management to the back end. With state management pushed to the back end, core service gains ability to execute multiple renku operation in sequence for those client (i.e batch execution). With the ability to batch multiple operations behind one interface, we could achieve simpler interactions with stateless clients. Hence the idea to add "porcelain" to the core service.

To understand better how service is designed, we should look at its renku operations through 3 main features - type of operation, possibility of remote execution, possibility of local execution. In short, types of operation are split into write or read operations (renku scoped operations). Local execution refers that we are executing renku operation on a project which is somewhere on the filesystem where the service is running. This could be in service defined cache or somewhere in user path. Remote execution means that user provides a remote address of a project and service needs to figure out the way to access that project, execute renku operations and sync it back to the remote.

Some of the work to support stateless clients has already been made, but mostly to support remote execution for renku **read** operations. The following table shows overview of current endpoints implementing renku operations:

Endpoint | Type of operations | Remote execution | Cache execution | Note
-- | -- | -- | -- | --
/datasets.create | Write | No | Yes |  
/datasets.add | Write | No | Yes |  
/datasets.remove | Write | No | Yes |  
/datasets.import | Write | No | Yes |  
/datasets.edit | Write | No | Yes |  
/datasets.unlink | Write | No | Yes |  
/cache.migrate | Write | No | Yes | Once remote execution is supported this endpoint should be renamed to /project.migrate
/datasets.files_list | Read | Yes | Yes |  
/datasets.list | Read | Yes | Yes |  
/cache.migrations_check | Read | Yes | Yes | Should probably be renamed to project.migrations_check
/graph.build | Read | Yes | Yes |  
/config.show | Read | Yes | Yes |
/config.set | Write | No | Yes |

As we can see, all **read** operations already fully support remote execution. Reason for this is because those read operations are relatively fast, cheap and require immediate response to the client. Write operations  on the other hand are more complex, slower and time of response doesn't matter, therefore should be implemented with asynchronous execution (life cycles for those executions is TBD). This gives us ability to implement them with batch execution in mind down the road.

**Stateless Write Operations (i.e. Porcelain) / Remote Execution**
- Higher level API for stateless clients which would like to focus on ease of use
- All operations executed in the background and status monitored through `/jobs` 
- Supports batched operations
	- (current) simple renku lifecycle: project_clone, migration_check, migrate, renku_op, sync
	- (future) renku workflows & dataset interactions: project_clone, migration_check, migrate, dataset import, workflow import, execute workflow with a dataset, sync
- Always sync with remote
- Manage entire operation lifecycle in the background
	- error management 
	- out of sync issues
	- push issues

**State-full Write Operations (i.e. Atomic) / Cache Execution**
- Lower level API with all of control  and flexibility needed
- Operate only on top of the cache or locally cloned repositories
- Used for implementation of renku clients in different languages such as R and Scala
- Execute only one operation (client is responsible for the setup of the operation)
- Decoupled operations
	- Don't sync with remote after each operation executed (needed for offline mode)

**TLDR;**
1. Adding remote execution to write endpoints removes the need to manage cache for stateless clients.
2. Keeping cache management enables state-full clients easier construction of batch operations and total control of renku operations (additional languages, offline support)
	- Cache management should be expanded to support additional operations such as sync and checkout 
3. (optional) API could be split between state-full  and stateless clients
	- `/datasets.create` would become `/datasets.create` for stateless and `/cache.datasets.create` for state-full.

**Comment by Panaetius** [comment]
The "Remote Execution" for write would work the same as it does now, i.e. clone the repo on each request without using the cache?

I'm a bit torn on if we want to still use the cache here, but manage it for the stateless client for performance reasons. On most projects, it probably doesn't make a big difference. On the budyko project, a clone (with skip_smudge) takes 24s. A pull takes 0.24 seconds, so it's a factor of 100 faster. Accessing a page in the UI performs a couple of things, checking migrations, listing datasets, potentially other. So this means 24s (for getting the project into cache + subsequent actions) vs. 48s or 72s (for multiple remote requests) page load times for that project.

**Comment by jsam** [comment]
That's a great point.  I see this problem as 2 fold, one is related to write operations and the other is related to adding improvements to existing read operations.

**Write**

The way I see implementation of remote execution for write operations is in the same way as a state-full client would behave. This means that those jobs would be cache aware and manage project state internally. Reason why I mentioned that "(life cycles for those executions is TBD)" is because we need to define clearly what this means, cause repository management goes beyond simple pull and tracking all the possible states we want to manage and should be mapped out. This includes management of sub-modules, out of sync issues and improvements to syncing with remotes.

**Read**

On the other hand, read operations have requirement of immediately returning the response. Here I see 3 possible improvements we could make and might be out of scope in here. Those are:
     

1. Add support to core service for renkulab awareness. 
          * This includes access to the KG and Gitlab. 
          * This support should be only activated in case of deployment to renkulab. 
          * This could also be a separate service which acts as a proxy between core service and entire renkulab
          * Build an index of existing repositories in renkulab 

2. Change implementation of remote reads in such a way that they only download needed metadata files and build the responses from there. This change would require edits on the core API for us to support this better. 
3. Make remote read operations work with cached repositories and manage repositories in cache accordingly

Now all 3 approaches have pros and cons. (1) would have fastest response times and we would give us ability to push updates to the UI clients, on the downside it might be harder to implement then the rest. (2) should be most straightforward to implement once we have support from the core module, (3) only pros I see here is that we could reuse design from the write operations, but we still have a problem that first request which would clone the project still be slow.
